Enable `mypy` and `isort` in pre-commit & CI

ml-evs opened the pull request:
This PR fixes many broken/dodgy type hints in the code, and a few associated bugs/undefined behaviours.

- [x] Enable mypy in CI
- [x] Enable and apply isort
- [x] Remove simplejson (not really needed since we dropped 3.7 support)
- [x] Update pre-commit
- [x] Add linters and fixers to dev requirements

JPBergsma commented:
In principle, this is a good idea.
I however think that we should solve all the warnings mypy gives (or silence them with # type: ignore if they are false alarms) before we merge this. 
Now it gives hundreds of warnings, I won't notice if my PR will add one extra warning, so I do not think it would be very useful at this point.

ml-evs commented:
Sorry for the kinda monstrous PR @JPBergsma, I think I will split it into two PRs with isort and then mypy. I didn't have to change any tests from my mypy changes so hopefully it should all just work...

JPBergsma commented:
For me, it is not necessary to split in two. I think both changes are quite orthogonal, so I can review both of them at the same time.

ml-evs commented:
> For me, it is not necessary to split in two. I think both changes are quite orthogonal, so I can review both of them at the same time.

Okay great, I guess my suggestion is to ignore the changes from [4782f5e](https://github.com/Materials-Consortia/optimade-python-tools/pull/1346/commits/4782f5e3cc5db93c2bcf9b9e09896fe5dd961590) and just check mypy first. The isort changes do not really need reviewing.

Once you are happy with it I will rebase this PR into a few commits so that we have the separation going forward.

ml-evs commented:
(We can also merge your other PRs before this one!)

JPBergsma commented:
I am still having some problems with PR#1304. 

The docker image setup seems to fail because of `yaml` missing. 
which is strange as this PR did not change anything about the dependancies.

And the tests still fail for the elastic search backend. 
So now I am trying to set up elastic locally to see why the tests are failing.

I tried to set up the docker container according to the installation instructions, but then I get an error that optimade can't connect to elastic search: "ERROR tests/server/middleware/test_api_hint.py - elasticsearch.exceptions.ConnectionError: ConnectionError(<urllib3.connection.HTTPConnection object at 0x7f4e5cdb62b0>: Failed to establish a new connection: [Errno 111] Connection ..." 

So I guess I still have some more work to do before the pr is merge ready.

JPBergsma commented on optimade/adapters/structures/aiida.py:
```diff
@@ -45,13 +44,13 @@ def get_aiida_structure_data(optimade_structure: OptimadeStructure) -> Structure
     species: Optional[List[OptimadeStructureSpecies]] = attributes.species
     if not species:
-        species = species_from_species_at_sites(attributes.species_at_sites)
+        species = species_from_species_at_sites(attributes.species_at_sites)  # type: ignore[arg-type]
```
The type of the argument in species_from_species_at_sites is List[str] and the type of species_at_sites is List[str] so the `#type: ignore` should not be necessary here. (I do not have Aiida so I could not test this.)

ml-evs replied:
It's `Optional[List[str]]` which is why mypy flags it up:

```
optimade/adapters/structures/aiida.py:53: error: Argument 1 to "species_from_species_at_sites" has incompatible type "Optional[List[str]]"; expected "List[str]"  [arg-type]
```

JPBergsma replied:
Thanks for the clarification. I did not know mypy was that strict.

JPBergsma commented on requirements-dev.txt:
```diff
@@ -1,9 +1,14 @@
 invoke==1.7.3
+isort==5.10.1
 jsondiff==2.0.0
+mypy==0.981
```
A newer version of mypy is out.
```suggestion
mypy==0.991
```

JPBergsma commented on tests/server/test_server_validation.py:
```diff
@@ -95,10 +95,7 @@ def test_versioned_base_urls(client, index_client, server: str):
-        import simplejson as json
-    except ImportError:
-        import json
+    import json
```
We already import json at the start of the file, so I do not see why we need to do this here again.
```suggestion
```

JPBergsma commented on tests/server/test_server_validation.py:
```diff
@@ -132,13 +129,10 @@ def test_meta_schema_value_obeys_index(client, index_client, server: str):
-        import simplejson as json
-    except ImportError:
-        import json
+    import json
```
```suggestion
```

ml-evs commented:
Just had to do a bit of a messy rebase after we merged that last PR, hope that doesn't mess up your review, but at least now the commits should each contain logical diffs

JPBergsma commented:
When I run mypy on my machine, I still get two errors.
1. optimade/adapters/structures/utils.py:223: error: Need type annotation for "Y"  [var-annotated]
    Replacing line 223 with     `Y: np.ndarray = np.cross(Z, X)` should solve this warning.
2. optimade/server/exception_handlers.py:231: error: List item 3 has incompatible type "Tuple[Type[VisitError], Callable[[Any, VisitError], JSONAPIResponse]]"; expected "Tuple[Type[Exception], Callable[[Any, Exception], JSONAPIResponse]]"  [list-item]
    
    I think black caused this line to be broken up so you need a second: `# type: ignore[list-item] # not entirely sure why this entry triggers mypy`

JPBergsma commented on optimade/adapters/structures/utils.py:
```diff
@@ -56,7 +58,7 @@ def scaled_cell(
         vector = np.cross(cell[(i + 1) % 3], cell[(i + 2) % 3]) / volume
         scale.append(tuple(vector))
-    return tuple(scale)
+    return tuple(scale)  # type: ignore[return-value]
```
The real problem here is that you cast the vector into a tuple. 
If you change line 60 to: `scale.append(Vector3D(vector))` you should not need `# type: ignore[return-value`

JPBergsma replied:
When I use pre-commit run mypy --all-files 
I still get an error here so the # type: ignore[return-value] should remain here.

JPBergsma commented on optimade/client/client.py:
```diff
@@ -93,7 +91,7 @@ class OptimadeClient:
     def __init__(
         self,
-        base_urls: Union[None, AnyUrl, List[AnyUrl]] = None,
+        base_urls: Optional[Union[str, List[str]]] = None,
```
On line 58 you defined base_urls as an Iterable yet here you define it as a list.
This is why mypy gives an error on lines 115 and 121
```suggestion
        base_urls: Optional[Union[str, Iterable[str]]] = None,
```

JPBergsma commented on optimade/filterparser/lark_parser.py:
```diff
@@ -30,13 +30,13 @@ def get_versions() -> Dict[Tuple[int, int, int], Dict[str, str]]:
     for filename in Path(__file__).parent.joinpath("../grammar").glob("*.lark"):
         tags = filename.stem.lstrip("v").split(".")
-        version = tuple(map(int, tags[:3]))
+        version = tuple(map(int, tags[:3]))  # ignore: type[index]
```
I think it should be # type: ignore[index]
But since mypy does not give any warning, I think we can remove it.
```suggestion
        version = tuple(map(int, tags[:3]))
```

ml-evs commented:
> When I run mypy on my machine, I still get two errors.
> 
>     1. optimade/adapters/structures/utils.py:223: error: Need type annotation for "Y"  [var-annotated]
>        Replacing line 223 with     `Y: np.ndarray = np.cross(Z, X)` should solve this warning.
> 
>     2. optimade/server/exception_handlers.py:231: error: List item 3 has incompatible type "Tuple[Type[VisitError], Callable[[Any, VisitError], JSONAPIResponse]]"; expected "Tuple[Type[Exception], Callable[[Any, Exception], JSONAPIResponse]]"  [list-item]

How have you configured your mypy? I would only trust the `pre-commit` environment to be consistent across different people (e.g., it depends what Python version you are actually invoking it from, and which add-on packages you have installed).

JPBergsma commented on optimade/models/responses.py:
```diff
@@ -67,17 +66,17 @@ class InfoResponse(Success):
 class EntryResponseOne(Success):
-    data: Union[EntryResource, Dict[str, Any], None] = Field(...)
-    included: Optional[Union[List[EntryResource], List[Dict[str, Any]]]] = Field(
+    data: Union[EntryResource, Dict[str, Any], None] = Field(...)  # type: ignore[assignment]
```
If you add Dict[str, Any] and List[Dict[str, Any]] to the possible data types of the data field in Response class in jsonapi.py
the ignore messages can be removed from lines 69 and 76

Similarly if you replace 366 in jsonapi.py with `included: Optional[Union[List[Resource], List[Dict[str, Any]]]] = StrictField(`
you can remove the type ignores from lines 70 and 79

ml-evs replied:
I tried this, but as `ErrorResponse` also uses `Response` (which then breaks the OpenAPI schema again), I think its better to be strict at the top-level then excuse ourselves where necessary.

JPBergsma commented on optimade/models/structures.py:
```diff
@@ -39,8 +39,8 @@
 
-Vector3D = conlist(float, min_items=3, max_items=3)
-Vector3D_unknown = conlist(Union[float, None], min_items=3, max_items=3)
+Vector3D = conlist(float, min_items=3, max_items=3)  # type: ignore[valid-type]
+Vector3D_unknown = conlist(Union[float, None], min_items=3, max_items=3)  # type: ignore[valid-type]
```
```suggestion
Vector3D = conlist(float, min_items=3, max_items=3)
Vector3D_unknown = conlist(Union[float, None], min_items=3, max_items=3)
```
On my machine it is not necessary to add type ignore here.

JPBergsma commented on optimade/server/entry_collections/entry_collections.py:
```diff
@@ -89,10 +91,10 @@ def __init__(
         self.provider_fields = [
             field if isinstance(field, str) else field["name"]
-            for field in CONFIG.provider_fields.get(resource_mapper.ENDPOINT, [])
+            for field in CONFIG.provider_fields.get(resource_mapper.ENDPOINT, [])  # type: ignore[call-overload]
```
```suggestion
            for field in CONFIG.provider_fields.get(resource_mapper.ENDPOINT, [])
```
I did not need to add #type ignore here.

JPBergsma commented on optimade/server/entry_collections/entry_collections.py:
```diff
@@ -293,7 +301,7 @@ def handle_query_params(
         if getattr(params, "filter", False):
             cursor_kwargs["filter"] = self.transformer.transform(
-                self.parser.parse(params.filter)
+                self.parser.parse(params.filter)  # type: ignore[union-attr]
```
I think it would be cleaner to add "filter" to the SingleEntryQueryParams class. 
We do use the filter to select a single entry based on the id. This may also allow us to make the code for both single and many responses more similar.
But I am not sure how much work it would be. So perhaps this can be done in another PR.

ml-evs replied:
It's not too much work, but it will break the OpenAPI schema and will present `filter` as a valid parameter for single entry endpoints (even though we do use the same machinery for it internally). I think its best we leave this one as is; the ignore is just because `mypy` isn't smart enough to resolve dynamic `getattr` guards.

JPBergsma commented:
> How have you configured your mypy? I would only trust the pre-commit environment to be consistent across different people (e.g., it depends what Python version you are actually invoking it from, and which add-on packages you have installed).
>
> Do you get the same errors if you do
>
> pre-commit run mypy --all-files?

With `pre-commit run mypy --all-files`
I still get the error on line 223 in optimade/adapters/structures/utils.py
The error on optimade/server/exception_handlers.py:231 however does disappear.

JPBergsma commented on optimade/server/entry_collections/entry_collections.py:
```diff
@@ -372,7 +380,7 @@ def parse_sort_params(self, sort_params: str) -> Tuple[Tuple[str, int]]:
 
         """
-        sort_spec = []
+        sort_spec: List[Tuple[str, int]] = []
```
Tuple should be changed to List in the docstring.

JPBergsma commented on optimade/server/logger.py:
```diff
@@ -1,9 +1,9 @@
 """Logging to both file and terminal"""
 import logging
+import logging.handlers
```
Why do you need to import logging.handlers? I think it will be imported when you import logging.

ml-evs replied:
This one isn't a mypy thing, but an annoying Python gotcha to do with automatic imports/global state: https://bugs.python.org/issue42842

ml-evs replied:
We were probably getting away with it as actually initializing the loggers before trying to access `logging.handlers` (which should then do the import behind the scenes) but probably better to do it this way to be explicit (hence the warning).

JPBergsma commented on optimade/server/mappers/entries.py:
```diff
@@ -179,7 +173,7 @@ def all_length_aliases(cls) -> Iterable[Tuple[str, str]]:
 
         return cls.LENGTH_ALIASES + tuple(
-            CONFIG.length_aliases.get(cls.ENDPOINT, {}).items()
+            CONFIG.length_aliases.get(cls.ENDPOINT, {}).items()  # type: ignore[call-overload]
```
```suggestion
            CONFIG.length_aliases.get(cls.ENDPOINT, {}).items()
```
For me, this "ignore" is also unnecessary.

ml-evs commented:
> With `pre-commit run mypy --all-files` I still get the error on line 223 in optimade/adapters/structures/utils.py The error on optimade/server/exception_handlers.py:231 however does disappear.
> 
> ps. Nearly all the other changes I have suggested this far remain valid. It is strange, we both get such different outcomes from mypy. Do you want me to continue reviewing the code or do you first want to decide what is causing the differences ?

Sorry, I missed this amongst the other notifications. I'll go through your existing comments and see if they make a difference for me (thanks for doing this so methodically). There's a chance that I added many of the ignores on the previous version of mypy (I think you had a comment above that updated it?) though I'd be surprised if they make that huge a difference.

ml-evs commented on optimade/filterparser/lark_parser.py:
```diff
@@ -30,13 +30,13 @@ def get_versions() -> Dict[Tuple[int, int, int], Dict[str, str]]:
 
     """
-    dct = defaultdict(dict)
+    dct: Dict[Tuple[int, int, int], Dict[str, Path]] = defaultdict(dict)
```
There wasn't huge value to using `defaultdict` here so I just refactored it out, no more type ignores needed

ml-evs commented:
> When I run mypy on my machine, I still get two errors.
> 
>     1. optimade/adapters/structures/utils.py:223: error: Need type annotation for "Y"  [var-annotated]
>        Replacing line 223 with     `Y: np.ndarray = np.cross(Z, X)` should solve this warning.
> 
>     2. optimade/server/exception_handlers.py:231: error: List item 3 has incompatible type "Tuple[Type[VisitError], Callable[[Any, VisitError], JSONAPIResponse]]"; expected "Tuple[Type[Exception], Callable[[Any, Exception], JSONAPIResponse]]"  [list-item]
>        I think black caused this line to be broken up so you need a second: `# type: ignore[list-item] # not entirely sure why this entry triggers mypy`

I can't reproduce these particular errors, but I've just addressed all your comments.

JPBergsma commented on optimade/filterparser/lark_parser.py:
```diff
@@ -96,7 +99,7 @@ def parse(self, filter_: str) -> Tree:
         try:
             self.tree = self.lark.parse(filter_)
-            self.filter = filter_
+            self.filter = filter_  # type: ignore[assignment]
```
For me   # type: ignore[assignment] is not needed anymore after you defined the type of self.filter on line 85

JPBergsma commented on optimade/server/middleware.py:
```diff
@@ -197,6 +192,8 @@ def handle_api_hint(api_hint: List[str]) -> Union[None, str]:
             # Go to /v<MAJOR>, since this should point to the latest available
             return BASE_URL_PREFIXES["major"]
 
+        return None
```
I do not see the point of adding a return None here. 
There does not seem to be a way to reach it.
And the value BASE_URL_PREFIXES["major"] seems to be a reasonable fallback value. 
So, could you not remove the last if statement instead?  
Ideally, you would also give a warning that a different API version has been used to handle the request.

ml-evs replied:
Agree on removing `return None` as it should be unreachable, this is one where I've just tried to placate mypy.

Instead I've reorganized it to be (shorthand)

```
if major_api_hint <= major_implementation:
   return hint

raise exc
```

I don't think we want to change the fallback value though (i.e., we should still raise an error if someone hints for v2).

JPBergsma replied:
That is also a good solution.

JPBergsma commented on optimade/server/middleware.py:
```diff
@@ -387,7 +386,7 @@ def showwarning(
                     # and the import machinery don't work anymore
                     line = None
-                    linecache = None
+                    linecache = None  # type: ignore[assignment]
```
Why do you set `linecache` to None here?  It does not seem to be used afterwords.
```suggestion
```

ml-evs replied:
The middleware warnings stuff is a bit fiddly and requires a few hacks like this. `linecache` is the name of a module that is otherwise not imported (see the comment section).

We shouldn't change this without good reason, hence why I just added the ignore

JPBergsma replied:
In the function you took this code from `linecache is still used later on in the function. So it makes sense to set it to `None` In our case we do not use the value of `linecache` later on so this kine can be removed.
In fact we could also replace the line above (`line = None`) with pass as the value of line is already `None`.

ml-evs replied:
Probably, but I don't think I want to do that in this PR. I didn't write this stuff initially (assume it was Casper) and there is limited benefit to changing it. The fact we are using private interfaces with the Python warnings library and linecache itself should indicate that this is a hacky bit of code!

ml-evs commented:
Thanks for the really thorough review @JPBergsma -- I think for the sake of our time, we can consider this PR ready... unless there's any objections I'll merge later today.

JPBergsma commented:
If you want, you can merge it. I still have not finished reviewing all the files yet, so I may still suggest some changes in a later PR.

ml-evs commented:
> If you want, you can merge it. I still have not finished reviewing all the files yet, so I may still suggest some changes in a later PR.

Okay, happy to hold off if you want to keep going.

JPBergsma commented on optimade/server/routers/utils.py:
```diff
@@ -116,7 +115,7 @@ def handle_response_fields(
     new_results = []
     while results:
-        new_entry = results.pop(0).dict(exclude_unset=True, by_alias=True)
+        new_entry = results.pop(0).dict(exclude_unset=True, by_alias=True)  # type: ignore[union-attr]
```
For me this  # type: ignore[union-attr] is also not neccesary.

JPBergsma commented on optimade/server/routers/utils.py:
```diff
@@ -252,21 +251,25 @@ def get_entries(
-    if fields or include_fields:
-        results = handle_response_fields(results, fields, include_fields)
+    if results is not None and (fields or include_fields):
+        results = handle_response_fields(results, fields, include_fields)  # type: ignore[assignment]
```
You need a "type ignore" here because handle_response_fields returns a list of dicts while results previously was a list of entry resources.
Changing the name of `results` on line 272 and  276 to something like "results_dicts" will prevent this error.

ml-evs replied:
I'm confused, there is already a `# type: ignore[assignment]` here?

JPBergsma replied:
Yes, but it can be removed if you rename `results` on lines 272 and 276 to something other than results.

ml-evs replied:
I think in a few cases the code is clearer if the ignores are kept in, rather than making additional temporary variables, so I've left this one in.

ml-evs commented:
I've just tried adding the `--warn_unused_ignores` config option to mypy and it doesn't even seem self-consistent. When I remove the ignores in the base adapters (after it says they are not needed), the same invocation then fails. Is this the same way you have been finding which ignores are not needed?

JPBergsma commented:
I mostly try to understand why the ignore statement is needed. 
If it is not clear, I remove it to see what kind of error mypy gives.
Sometimes I do not get an error, and in that case I make a remark here that the ignore statement is not needed.

JPBergsma commented on optimade/validator/utils.py:
```diff
@@ -365,15 +361,15 @@ def wrapper(
                         message += traceback.split("\n")
 
-                message = "\n".join(message)
+                message = "\n".join(message)  # type: ignore[assignment]
```
You get an error here because you store a string in a variable `message` that has the type list.
One way to solve this would be to replace `message` on line 384 with  `"\n".join(message)` and delete line 372.

I d

JPBergsma commented on tests/server/utils.py:
```diff
@@ -54,24 +49,11 @@ def __init__(
         self.version = version
 
-    def request(  # pylint: disable=too-many-locals
+    def request(  # type: ignore[override] # pylint: disable=too-many-locals
```
For me, `type: ignore[override]` is not needed here.

ml-evs commented:
Thanks @JPBergsma. I've addressed your final comments and will merge once the tests pass.

JPBergsma commented on optimade/validator/utils.py:
```diff
@@ -365,16 +361,15 @@ def wrapper(
 
-                message = "\n".join(message)
-
+                failure_type = None
```
I am not so happy with this line. The value of failure_type will be overwritten later on, so this assignment looks rather strange.
Is it such a problem to define the type on what is now line 377?

ml-evs replied:
Not a problem, but not of any importance, I don't think. I'd rather have the default value outside of the clause rather than a type-hint inside the clause. Not everything has to be explicitly typed (and we shouldn't let mypy dictate how we code beyond any real errors, especially as it changes between versions).

JPBergsma reviewed (approved):
LGTM
